src/display/annotation_layer.js

```diff
@@ -2366,18 +2393,70 @@ class AnnotationLayer {
    * @memberof AnnotationLayer
    */
   static update(parameters) {
-    const transform = `matrix(${parameters.viewport.transform.join(",")})`;
+    const { page, viewport, annotations, annotationCanvasMap, div } =
+      parameters;
+    const transform = viewport.transform;
+    const matrix = `matrix(${transform.join(",")})`;
+
+    let scale, ownMatrix;
+    if (annotations.some(annotation => annotation.hasOwnCanvas)) {
+      scale = Math.abs(transform[0] || transform[1]);
+      const ownTransform = transform.slice();
+      for (let i = 0; i < 4; i++) {
+        ownTransform[i] = Math.sign(ownTransform[i]);
+      }
+      ownMatrix = `matrix(${ownTransform.join(",")})`;
+    }
+
     for (const data of parameters.annotations) {
-      const elements = parameters.div.querySelectorAll(
+      const elements = div.querySelectorAll(
         `[data-annotation-id="${data.id}"]`
       );
       if (elements) {
         for (const element of elements) {
-          element.style.transform = transform;
+          if (data.hasOwnCanvas) {
+            const rect = Util.normalizeRect([
+              data.rect[0],
+              page.view[3] - data.rect[1] + page.view[1],
+              data.rect[2],
+              page.view[3] - data.rect[3] + page.view[1],
+            ]);
+
+            const left = rect[0] * scale;
+            const top = rect[1] * scale;
+            element.style.left = `${left}px`;
+            element.style.top = `${top}px`;
+            element.style.transformOrigin = `${-left}px ${-top}px`;
+            element.style.transform = ownMatrix;
+          } else {
+            element.style.transform = matrix;
+          }
         }
       }
     }
-    parameters.div.hidden = false;
+
+    this.setAnnotationCanvasMap(div, annotationCanvasMap);
+    div.hidden = false;
+  }
+
+  static setAnnotationCanvasMap(div, annotationCanvasMap) {
```

Comment:
Let's make this properly private, since is shouldn't be called directly; i.e. `static #setAnnotationCanvasMap` instead.